### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jPeachPBA**
Refactored code in telmos_main and telmos_goods so that repeated blocks are in a separate function 
Split telmos_main into smaller functions that represent the process more clearly

Cleaned up code relating to threads in gui files
Removed Student Factors.txt  - now directly within the code
Revised Readme.md

### CODE REVIEW

**Comment by asongtoruin** [comment]
Can you explain what's going on in [this commit](https://github.com/TransportScotland/tmfs18-trip-end-model/pull/4/commits/79d22ac2405723d1d101b5f45c54c1f8f07e051b) a little more? Ideally it would be multiple separate commits, as there seem to be unconnected things going on.

**Comment by jPeachPBA** [comment]
In [this commit](https://github.com/TransportScotland/tmfs18-trip-end-model/pull/4/commits/79d22ac2405723d1d101b5f45c54c1f8f07e051b) the student factors file was removed, the readme updated and two flags were removed - do_debug, and do_output - which were not necessary anymore.
The options to select the flags was removed from gui.py and any references to them in the other files were removed (or moved out of an "if" block). This made up most of the changes.
hgv_count was also renamed to zone_count in the telmos_goods.py

File: telmos_main.py [hunk 1]
```diff
@@ -12,22 +12,10 @@
...
-
-
-def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
-                base_year, base_id, base_scenario, is_rebasing_run=True,
-                do_output=False, debug=True, log_func=print,
-                just_pivots=False):
+def read_trip_rates(factors_dir, just_pivots):
     '''
-    This is a conversion of the TMfS14 Visual Basic application 
+    Loads the production trip rate files into
```
Comment (asongtoruin) [inline telmos_main.py]: Docstring needs finishing

File: telmos_main.py [hunk 2]
```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
...
+    # Columns are: Work, Employment, Other, Education
+    attr_factors_array = np.zeros((planning_data.shape[0], 4),dtype="float32")
+    attr_factors_array[:,0] = planning_data[:,2] * attraction_trip_rates[0, 0]
+    attr_factors_array[:,1] = planning_data[:,[1,3,4,5,6,7,8]].sum(axis=1) * attraction_trip_rates[2,1]
+    attr_factors_array[:,2] = (planning_data[:,[4,8,1,3]] * 
+                      attraction_trip_rates[[6,7,1,12],[3,4,8,9]]).sum(axis=1)
+    attr_factors_array[:,3] = planning_data[:,7] * attraction_trip_rates[2,2]
+    # Fill zeros with ones
+    attr_factors_array[attr_factors_array == 0.0] = 1.0
+    return attr_factors_array
```
Comment (asongtoruin) [inline telmos_main.py]: If we're just resetting to 1 anyway, can we not just use `np.ones`?

```suggestion
    attr_factors_array = np.ones((planning_data.shape[0], 4),dtype="float32")
    attr_factors_array[:,0] = planning_data[:,2] * attraction_trip_rates[0, 0]
    attr_factors_array[:,1] = planning_data[:,[1,3,4,5,6,7,8]].sum(axis=1) * attraction_trip_rates[2,1]
    attr_factors_array[:,2] = (planning_data[:,[4,8,1,3]] * 
                      attraction_trip_rates[[6,7,1,12],[3,4,8,9]]).sum(axis=1)
    attr_factors_array[:,3] = planning_data[:,7] * attraction_trip_rates[2,2]
    return attr_factors_array
```

File: gui.py [hunk 3]
```diff
@@ -64,17 +65,19 @@ def init_widgets(self):
         # Directory information
         directory_frame = ttk.Frame(input_frame, borderwidth=3, relief=tk.GROOVE)
         directory_frame.pack()
-        delta_dir_tt = ("Contains paths - '{Base Scenario}/trfl{Base Year}"
-                        "{Base Scenario}.dat' and '{Future Scenario}/"
-                        "[tav_/tmfs/trfl]{Future Year}{Future Scenario}.dat'")
+        delta_dir_tt = ("Contains folders containing the TELMoS planning data csv files "
+                        "and goods dat files. Folders and files should be named according to the "
+                        "year and scenario code.")
```
Comment (asongtoruin) [inline gui.py]: Can we have an example of the required structure in here?

File: gui.py [hunk 4]
```diff
@@ -64,17 +65,19 @@ def init_widgets(self):
...
                                   pack_side="left", inter_pack_side="top",
                                   w=30, text_style="HEAD.TLabel",
                                   tool_tip_text=delta_dir_tt)
         delta_dir.add_directory()
-        tmfs_dir_tt = ("Contains paths - 'Runs/{Year}/Demand/{ID}/' for both "
-                       "base and demand scenarios")
+        tmfs_dir_tt = ("Contains the 'Factors' folder and the 'Runs' folder.\n"
+                       "'Factors' contains all internal files used by the model\n"
+                       "'Runs' contains the folder structure 'Year'/Demand/'ID' "
+                       "and should also include the base year files")
```
Comment (asongtoruin) [inline gui.py]: Can we have an example of the required structure in here?

File: telmos_addins.py [hunk 5]
```diff
@@ -81,14 +75,12 @@ def telmos_addins(delta_root, tmfs_root,
...
-            if do_output:
-                out_file = os.path.join(tmfs_root, "Runs", tel_year, 
-                                               "Demand", tel_id, filename)
-                matrix_to_odfile([x.round(3) for x in new_addin_array[f_key]], 
-                                  out_file, num_columns=3)
-                log_func("Saved matrix as %s" % str(out_file))
-                produced_files.append(out_file)
-                file_names.append(filename)
+            matrix_to_odfile([x.round(3) for x in new_addin_array[f_key]], 
+                              out_file, num_columns=3)
```
Comment (asongtoruin) [inline telmos_addins.py]: Seems very similar to line 64? Can we catch both instances with a line outside the `if` statement?

File: telmos_addins.py [hunk 6]
```diff
@@ -118,30 +110,12 @@ def telmos_addins(delta_root, tmfs_root,
...
-            c_total = np.loadtxt(c, skiprows=skiprows, delimiter=",").sum()
-            diffs[name] = (p_total - c_total) / c_total
-        return diffs
-                
-                
-                
-                
+        out_file = os.path.join(tmfs_root, "Runs", tel_year, "Demand",
+                                tel_id, filename.replace(".DAT", "TE.DAT"))
+        format_string = ["%d"] + ["%.3f" for x in range(te_array.shape[1]-1)]
```
Comment (asongtoruin) [inline telmos_addins.py]: Convention is to use `_` for throwaways

```suggestion
        format_string = ["%d"] + ["%.3f" for _ in range(te_array.shape[1]-1)]
```

File: telmos_goods.py [hunk 7]
```diff
@@ -11,13 +11,34 @@
...
+    lgv_array = pd.DataFrame(data[data[:,0]==2][:,1:])
+    #Print to separate files
+    pd.concat((hgv_array.loc[:,:1].astype("int16"),hgv_array.loc[:,2]),axis=1).to_csv(
+        hgv_output, index=False, header=False)
+    pd.concat((lgv_array.loc[:,:1].astype("int16"),lgv_array.loc[:,2]),axis=1).to_csv(
+        lgv_output, index=False, header=False)
+    
+    hgv_array = np.array(hgv_array.set_index([0,1]).unstack(-1))
+    lgv_array = np.array(lgv_array.set_index([0,1]).unstack(-1))
+    return (hgv_array, lgv_array)
```
Comment (asongtoruin) [inline telmos_goods.py]: This function is a bit messy, and doesn't really use the useful features of `pandas`. Here's a possible rewrite (probably would need to be in its own commit?) that is a little bit longer in terms of line numbers, but I think should be more readable?

```suggestion
def load_goods_data(goods_file, hgv_output, lgv_output):
    '''
    Loading function for TELMoS goods files
    Splits the data into LGV and HGV parts
    '''
    hgv_lines = []
    lgv_lines = []
    # Inconsistent format of input data - iterate through
    # lines, use indicator to check if HGV / LGV
    with open(goods_file, "r") as f:
        for line in f:
            split_line = line.split()
            indicator = split_line.pop(0)
            if indicator == "1":
                hgv_lines.append(split_line)
            elif indicator == "2":
                lgv_lines.append(split_line)
            else:
                continue
    
    # Frame should have 3 columns
    # TODO: in both cases, check all lines have the right number of entries.
    hgv_df = pd.DataFrame(hgv_lines, columns=["I", "J", "V"], dtype="float64")
    lgv_df = pd.DataFrame(lgv_lines, columns=["I", "J", "V"], dtype="float64")
    
    # Rewrap zone numbers as integers
    for df in (hgv_df, lgv_df):
        df["I"] = df["I"].astype(int)
        df["J"] = df["J"].astype(int)
    
    hgv_df.to_csv(hgv_output, index=False, header=False)
    lgv_df.to_csv(lgv_output, index=False, header=False)
    
    hgv_array = hgv_df.pivot_table(index="I", columns="J").to_numpy()
    lgv_array = lgv_df.pivot_table(index="I", columns="J").to_numpy()
    
    return hgv_array, lgv_array
```

File: telmos_goods.py [hunk 8]
```diff
@@ -179,80 +156,32 @@ def telmos_goods(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
...
-        # Print these files as stack
-        if do_output:
-            df = pd.DataFrame(new_forecast_array[f_key]).stack().reset_index()
-            df.loc[:,:"level_1"] = df.loc[:,:"level_1"] + 1
-            df.to_csv(os.path.join(tel_filebase, filename), index=False, 
-                      header=False, float_format="%.3f")
-            log_func("Goods file saved to %s" % str(os.path.join(tel_filebase, filename)))
+        # Print these files stacked
+        df = pd.DataFrame(new_forecast_array[f_key]).stack().reset_index()
+        df.loc[:,:"level_1"] = df.loc[:,:"level_1"] + 1
```
Comment (asongtoruin) [inline telmos_goods.py]: This is a bit gross. Might help to make it more explicit what's happening in the `DataFrame`:
```suggestion
        # Get matrix in "long" form to write as CSV 
        df = pd.DataFrame(new_forecast_array[f_key]).stack().reset_index()
        df.columns = ['I', 'J', 'V']
        
        # Change zone numbers from 0-based to 1-based
        df['I'] += 1
        df['J'] += 1
```

File: telmos_main.py [hunk 9]
```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
     suffixes = range(3, 9)
     
     sr_array = []
-    all_names = []
     for suffix in suffixes:
         data = []
         names = []
```
Comment (asongtoruin) [inline telmos_main.py]: This isn't used - can we remove it? See also line 34

File: telmos_main.py [hunk 10]
```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
...
+    for t_file, c_file in zip(tod_files, cte_files):
+        # TMfS14 had a long-distance module that required some CTE/TOD files to have
+        # _All appended to the end. This can now be removed if needed.
+        if (os.path.exists(os.path.join(file_base, t_file)) and
+               os.path.exists(os.path.join(file_base, c_file))):
+            t_file_name = t_file
+            c_file_name = c_file
+        else:
+            t_file_name = t_file.replace("_ALL","")
+            c_file_name = c_file.replace("_ALL","")
```
Comment (asongtoruin) [inline telmos_main.py]: Current set-up isn't flexible for if one file ends in `_ALL` and the other doesn't. How about:

```suggestion
        t_file_name = (
            t_file if os.path.isfile(os.path.join(file_base, t_file))
            else t_file.upper().replace("_ALL", "")
        )
        c_file_name = (
            c_file if os.path.isfile(os.path.join(file_base, c_file))
            else c_file.upper().replace("_ALL", "")
        )
```

This also:

- ensures our `_ALL` is properly replaced, by converting the filename to uppercase
- uses `os.path.isfile` rather than `os.path.exists`, so it can't be accidentally fooled by a folder with the name we're looking for.

File: telmos_main.py [hunk 11]
```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
...
+            if j == cb:
+                cb += 4
+            if j == cc:
+                cc += 4
+            if j == cd:
+                cd += 4
+
+            prod_factor_array[i, j] = v
+                
+    prod_factor_array[prod_factor_array == 0] = 1
```
Comment (asongtoruin) [inline telmos_main.py]: Less sure this time - but would it be possible to use `np.ones` rather than `np.zeros` if we're just resetting to 1 at the end?

File: telmos_main.py [hunk 12]
```diff
@@ -36,56 +24,253 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
...
     # Load in the student factors and attraction factors separately
-    filenames = ["Student Factors.txt", "Attraction Factors.txt"]
-    student_factors = np.loadtxt(os.path.join(factors_base, filenames[0]))
-    attraction_factors = np.loadtxt(os.path.join(factors_base, filenames[1]))
+    attraction_file = "Attraction Factors.txt"
+    attraction_factors = np.loadtxt(os.path.join(factors_base, attraction_file))
+        
+    # [Male student factor, Female student factor]
+    # Used to factor the productions down
+    student_factors = np.array([0.2794, 0.2453])
```
Comment (asongtoruin) [inline telmos_main.py]: These shouldn't be an array. There's only two values, so having them in an array just makes their actual meaning more opaque.

I think one option might be to make them global factors that indicate what they're used for, at the top of the file, e.g.

```python
import numpy as np
import pandas as pd


MALE_STUDENT_FACTOR = 0.2794
FEMALE_STUDENT_FACTOR = 0.2453

def read_trip_rates(factors_dir, just_pivots):
# ...
```

**Review by asongtoruin — CHANGES_REQUESTED** [review]
Some feedback here - a few directly suggested changes, and a few more general comments. If you're going to add these changes, they should be split across several commits.

File: gui.py (hunk 3 above)
Comment (jPeachPBA) [reply gui.py]: I've added examples into the tooltips, although this does lead to them being quite long. Not sure if there is a better way of showing an example.

File: telmos_main.py (hunk 9 above)
Comment (jPeachPBA) [reply telmos_main.py]: Couldn't see this example in the file - there were other places where variables were not used and these have been removed.

File: gui.py (hunk 3 above)
Comment (asongtoruin) [reply gui.py]: Hmm. Could it possibly be a text box rather than a tooltip then? To be honest, I don't think it's particularly obvious that there is a tooltip, so if we're going to stick with that it might be better to have e.g. a question mark by each section that has the tooltip connected. Doesn't necessarily matter for this PR though, so maybe we mark it as resolved for now and discuss in a new Issue?

**Review by asongtoruin — APPROVED** [review]

**Review by asongtoruin — APPROVED** [review]

**Comment by asongtoruin** [comment]
This closes #2 and #3
